File: examples/c/03_transmission_spectrum.c
#include "common_example.h"

#include <stdlib.h>

int main(void)
{
    optics_output_init_viz("core_plugins");

    const double wl_tab[] = {200, 300, 400, 500, 750};
    const optics_complex nk_tab[] = {
        {2.1, 0.1}, {2.4, 0.3}, {2.3, 0.4}, {2.2, 0.4}, {2.2, 0.5}};
    const size_t n_tab = 5;
    const double depth = 300.0;

    const size_t N = 400;
    double lambda[N], T[N];
    for (size_t i = 0; i < N; ++i) {
        lambda[i] = 200.0 + (550.0 * i) / (N - 1);
        optics_layer_stack* stack = optics_layer_stack_create();
        optics_layer_stack_add_layer(stack, OL(1.0, 0.0, 0.0));
        optics_complex nk = interp_nk_quadratic(wl_tab, nk_tab, n_tab, lambda[i]);
        optics_layer_stack_add_layer(stack, (optics_layer){nk, depth});
        optics_layer_stack_add_layer(stack, OL(1.0, 0.0, 0.0));

        optics_complex dir[3];
        size_t n_dir = 3;
        optics_tmm_propagate_direction(stack, (optics_complex){0.0, 0.0}, dir, &n_dir);
        optics_matrix2x2 tmm;
        tmm_last_matrix_s(stack, dir, n_dir, lambda[i], &tmm);
        optics_layer l0, l2;
        optics_layer_stack_get_layer(stack, 0, &l0);
        optics_layer_stack_get_layer(stack, 2, &l2);
        double r = 0, t = 0;
        optics_tmm_get_r_t_power_from_tmm_s(tmm, l0.nk, dir[0], l2.nk, dir[2], &r, &t);
        T[i] = t;
        optics_layer_stack_destroy(stack);
    }

    optics_output_plot_xy_or_csv("output/03_transmission_spectrum.csv", lambda, T, N,
        "Transmission at normal incidence", "Wavelength (nm)", "Fraction of power transmitted", "b-");

    optics_py_shutdown();
    return EXIT_SUCCESS;
}

File: examples/fortran/03_transmission_spectrum.f90
program ex03_transmission_spectrum
  use optics_toolkits
  use csv_output
  implicit none

  type(optics_layer_stack) :: stack
  type(optics_complex) :: dir(3), crao, nk_val
  type(optics_matrix2x2) :: tmm
  type(optics_layer) :: l0, l2
  real(8) :: wl_tab(5), lambda(400), tval(400), rp, tp
  type(optics_complex) :: nk_tab(5)
  real(8), parameter :: depth = 300.0d0
  integer :: n_dir, i

  wl_tab = (/200.0d0, 300.0d0, 400.0d0, 500.0d0, 750.0d0/)
  nk_tab(1) = oc(2.1d0, 0.1d0)
  nk_tab(2) = oc(2.4d0, 0.3d0)
  nk_tab(3) = oc(2.3d0, 0.4d0)
  nk_tab(4) = oc(2.2d0, 0.4d0)
  nk_tab(5) = oc(2.2d0, 0.5d0)

  call ensure_output_dir('output')

  do i = 1, 400
    lambda(i) = 200.0d0 + (550.0d0 * dble(i - 1)) / 399.0d0

    call stack%create()
    call stack%add_layer(ol(1.0d0, 0.0d0, 0.0d0))
    nk_val = interp_nk_quadratic(wl_tab, nk_tab, 5, lambda(i))
    call stack%add_layer(ol(nk_val%re, nk_val%im, depth))
    call stack%add_layer(ol(1.0d0, 0.0d0, 0.0d0))

    crao = oc(0.0d0, 0.0d0)
    n_dir = 3
    call tmm_propagate_direction(stack, crao, dir, n_dir)
    call tmm_last_matrix_s(stack, dir, n_dir, lambda(i), tmm)
    l0 = stack%get_layer(1)
    l2 = stack%get_layer(3)
    call tmm_get_r_t_power_from_tmm_s(tmm, l0%nk, dir(1), l2%nk, dir(3), rp, tp)
    tval(i) = tp
  end do

  call write_csv_2col('output/03_transmission_spectrum.csv', 'Wavelength (nm)', &
    'Fraction of power transmitted', lambda, tval, 400)
end program ex03_transmission_spectrum
